flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGateMetrics.java

```diff
@@ -0,0 +1,168 @@
+
+	private long lastTotal = -1;
+
+	private int lastMin = -1;
+
+	private int lastMax = -1;
+
+	private float lastAvg = -1.0f;
+
+	// ------------------------------------------------------------------------
+
+	private InputGateMetrics(SingleInputGate inputGate) {
+		this.inputGate = checkNotNull(inputGate);
+	}
+
+	// ------------------------------------------------------------------------
+
+	// these methods are package private to make access from the nested classes faster 
+
+	long refreshAndGetTotal() {
+		long total;
+		if ((total = lastTotal) == -1) {
+			refresh();
+			total = lastTotal;
+		}
+
+		lastTotal = -1;
+		return total;
+	}
+
+	int refreshAndGetMin() {
+		int min;
+		if ((min = lastMin) == -1) {
+			refresh();
+			min = lastMin;
+		}
+
+		lastMin = -1;
+		return min;
+	}
+
+	int refreshAndGetMax() {
+		int max;
+		if ((max = lastMax) == -1) {
+			refresh();
+			max = lastMax;
+		}
+
+		lastMax = -1;
+		return max;
+	}
+
+	float refreshAndGetAvg() {
+		float avg;
+		if ((avg = lastAvg) < 0.0f) {
+			refresh();
+			avg = lastAvg;
+		}
+
+		lastAvg = -1.0f;
+		return avg;
+	}
+
+	private void refresh() {
+		long total = 0;
+		int min = Integer.MAX_VALUE;
+		int max = 0;
+		int count = 0;
+
+		for (InputChannel channel : inputGate.getInputChannels().values()) {
+			if (channel.getClass() == RemoteInputChannel.class) {
+				RemoteInputChannel rc = (RemoteInputChannel) channel;
+
+				int size = rc.unsynchronizedGetNumberOfQueuedBuffers();
+				total += size;
+				min = Math.min(min, size);
+				max = Math.max(max, size);
+				count++;
+			}
+		}
+
+		this.lastTotal = total;
+		this.lastMin = min;
+		this.lastMax = max;
+		this.lastAvg = total / (float) count;
+	}
+
+	// ------------------------------------------------------------------------
+	//  Gauges to access the stats
+	// ------------------------------------------------------------------------
+
+	private Gauge<Long> getTotalQueueLenGauge() {
+		return new Gauge<Long>() {
+			@Override
+			public Long getValue() {
+				return refreshAndGetTotal();
+			}
+		};
+	}
+
+	private Gauge<Integer> getMinQueueLenGauge() {
+		return new Gauge<Integer>() {
+			@Override
+			public Integer getValue() {
+				return refreshAndGetMin();
+			}
+		};
+	}
+
+	private Gauge<Integer> getMaxQueueLenGauge() {
+		return new Gauge<Integer>() {
+			@Override
+			public Integer getValue() {
+				return refreshAndGetMax();
+			}
+		};
+	}
+
+	private Gauge<Float> getAvgQueueLenGauge() {
+		return new Gauge<Float>() {
+			@Override
+			public Float getValue() {
+				return refreshAndGetAvg();
+			}
+		};
+	}
+
+	// ------------------------------------------------------------------------
+	//  Static access
+	// ------------------------------------------------------------------------
+
+	public static void registerQueueLengthMetrics(MetricGroup group, SingleInputGate gate) {
+		InputGateMetrics metrics = new InputGateMetrics(gate);
+
+		group.gauge("total-queue-len", metrics.getTotalQueueLenGauge());
```

Comment:
These metric names aren't consistent with other metrics; they should be named ```totalQueueLen``` or similar.